clang/include/clang/Basic/Attr.td

```diff
@@ -1432,6 +1432,24 @@ def SYCLIntelMaxGlobalWorkDim : InheritableAttr {
   let SupportsNonconformingLambdaSyntax = 1;
 }
 
+def SYCLDeviceGlobal: InheritableAttr {
+  let Spellings = [GNU<"device_global">,
+                   CXX11<"__sycl_detail__", "device_global">];
+  let Subjects = SubjectList<[CXXRecord], ErrorDiag>;
+  let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
+  let Documentation = [SYCLDeviceGlobalAttrDocs];
+  let SimpleHandler = 1;
+}
+
+def SYCLGlobalVariableAllowed : InheritableAttr {
+  let Spellings = [GNU<"global_variable_allowed">,
+                   CXX11<"__sycl_detail__", "global_variable_allowed">];
+  let Subjects = SubjectList<[CXXRecord], ErrorDiag>;
+  let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
+  let Documentation = [SYCLGlobalVariableAllowedAttrDocs];
+  let SimpleHandler = 1;
+}
+
```

Comment:
Both attributes are doing same but only spellings are different. Can we combine them into one attribute? Is there a reason to have a separate attribute here? 
```
def SYCLDeviceGlobal: InheritableAttr {
  let Spellings = [GNU<"device_global">,
                           CXX11<"__sycl_detail__", "device_global">,
                           GNU<"global_variable_allowed">,
                           CXX11<"__sycl_detail__", "global_variable_allowed">];;
  let Subjects = SubjectList<[CXXRecord], ErrorDiag>;
  let Accessors = [Accessor<"isGlobalVariable", 
                           [GNU<"global_variable_allowed">,
                           CXX11<"__sycl_detail__", "global_variable_allowed">]>];
  let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
  let Documentation = [SYCLDeviceGlobalAttrDocs];
  let SimpleHandler = 1;
}
```
@Fznamznon, @AaronBallman, what is your thought? This will reduce duplicate codes.

Reply:
If I remember correctly two attributes are needed to allow non-const global variables in the device code for more types (not only device_global). I think @gmlueck can explain it in the best way.

Reply:
The `[[device_global]]` and `[[global_variable_allowed]]` attributes do not have the same semantic, so they should cause different behavior in the CFE.  From the [design document](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/DeviceGlobal.md):

> The `[[__sycl_detail__::global_variable_allowed]]` attribute informs the front-end that global variables of this type are allowed to be referenced in device code. By default, the front-end diagnoses an error if device code references a global variable unless the variable is `constexpr` or `const` and constant initialized. However, the presence of this attribute informs the front-end that variables of this type are an exception to this rule, so the front-end does not diagnose an error when device code references a device_global variable. This attribute could also be used by other types, so it is also not specific to the device_global class.

And:

> The last attribute `[[__sycl_detail__::device_global]]` controls error reporting for variables declared of this type. The device global extension specification places restrictions on where a device_global variable can be declared. Rather than have the front-end recognize the name of the device_global type, the front-end uses this attribute to know which restrictions to enforce for this type.

The restrictions mentioned in that description are the ones listed in the device global extension specification section titled "[Restrictions on creating device global objects](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_global.asciidoc#restrictions-on-creating-device-global-objects)".

Reply:
Thank you @gmlueck and @Fznamznon for the explanation.